code/modules/antagonists/heretic/heretic_antag.dm

```diff
@@ -84,33 +84,99 @@
 	LAZYNULL(sac_targets)
 	return ..()
 
+/datum/antagonist/heretic/proc/get_icon_of_knowledge(datum/heretic_knowledge/knowledge)
+	if(ispath(knowledge))
+		//if the argument is a typepath, we need to make a dummy so we can access some of its properties
+		knowledge = new knowledge()
+
+	//if the knowledge has a special icon, use that
+	if(!isnull(knowledge.research_tree_icon))
+		return knowledge.research_tree_icon
+
+	//if the knowledge is a transmutation, use the result
+	if(!isnull(knowledge.result_atoms) && knowledge.result_atoms.len)
+		var/atom/result_thing = knowledge.result_atoms[1]
+		return icon(result_thing.icon,result_thing.icon_state,frame=1)
+
+	//if the knowledge is a spell, use the spell's button
+	if(istype(knowledge,/datum/heretic_knowledge/spell))
+		var/datum/heretic_knowledge/spell/spell_knowledge = knowledge
+		var/datum/action/cooldown/spell/result_spell = spell_knowledge.spell_to_add
+		return icon(result_spell.button_icon,result_spell.button_icon_state,frame=1)
+
+	//if the knowledge is a summon, use the mob sprite
+	if(istype(knowledge,/datum/heretic_knowledge/summon))
+		var/datum/heretic_knowledge/summon/summon_knowledge = knowledge
+		var/mob/living/result_mob = summon_knowledge.mob_to_summon
+		return icon(result_mob.icon,result_mob.icon_state,frame=1,dir=SOUTH,moving=0)
+
+	//if the knowledge is an eldritch mark, use the mark sprite
+	if(istype(knowledge,/datum/heretic_knowledge/mark))
+		var/datum/heretic_knowledge/mark/mark_knowledge = knowledge
+		var/datum/status_effect/eldritch/mark_effect = mark_knowledge.mark_type
+		return icon(mark_effect.effect_icon,mark_effect.effect_icon_state,frame=1)
+
+	//if the knowledge is an ascension, use the achievement sprite
+	if(istype(knowledge,/datum/heretic_knowledge/ultimate))
+		var/datum/heretic_knowledge/ultimate/ascension_knowledge = knowledge
+		var/datum/award/achievement/misc/achievement = ascension_knowledge.ascension_achievement
+		if(!isnull(achievement))
+			return icon(achievement.icon,achievement.icon_state)
+
+	//if all else fails, use a default icon
+	return icon('icons/mob/actions/actions_ecult.dmi',"eye")
+
 /datum/antagonist/heretic/ui_data(mob/user)
 	var/list/data = list()
 
 	data["charges"] = knowledge_points
 	data["total_sacrifices"] = total_sacrifices
 	data["ascended"] = ascended
 
+	//setup tiers
+	var/list/tiers = list()
+	for(var/i in 0 to 10)
+		tiers += list(list("nodes"=list()))
```

Comment:
This should be a define rather than a magic number

Reply:
maybe expanding the list as we add the knowledge elements would be even better. I think I'll rewrite that part entirely